### ISSUE TITLE
Common developer-tools container with credentials handling

### ISSUE THREAD

**Issue by jeffmccune** [GoogleCloudPlatform/cloud-foundation-toolkit#255]
This ticket captures discussions and decisions regarding credential handling when running automated tests for CFT Terraform modules.  The scope of this design is local test development on a developer's workstation and also in CI.

Containers
===

For both local development and CI, a single `developer-tools` container is used.  This container includes all necessary dependencies to run the lint, unit and integration tests for a CFT terraform module.

Credentials
===

CFT terraform modules commonly use the following tools which need credentials to be initialized within the context of the running container.  Each tool has a slightly different way of handling credentials:

 1. terraform
 2. inspec
 3. gcloud
 4. gsutil

If additional tools a required as dependencies:

 1. Install the tool into the developer-tools image via the Dockerfile
 2. If the new tool does not respect one of the existing credentials mechanisms (e.g. GOOGLE_CREDENTIALS), then update the `init_credentials` function inside of `/usr/local/bin/task_helper_functions.sh` to initialize the credentials from SERVICE_ACCOUNT_JSON for the new tool.
 3. Build and release a new version of the developer-tools image
 4. Update CI pipelines to use the new version as necessary.

Design Decision
===

Credentials are to be passed into the container using a single environment variable, `SERVICE_ACCOUNT_JSON`.  This variable name is chosen for backwards compatibility with existing CI jobs and to avoid a naming conflict with `GOOGLE_CREDENTIALS`.  The value of `SERVICE_ACCOUNT_JSON` must be the actual JSON string content of a service account JSON key.  For example `SERVICE_ACCOUNT_JSON="$(< ~/.credentials/jmccune-devtools-0abcde01234.json`  The value must not be a filesystem path.

Filesystem Paths
===

Generally, speaking filesystems paths should be avoided when loading SA credentials.  The intent is to rely on automatic loading of credentials when possible so that jobs may be deployed with service accounts managed by the platform.  For example, consider a VM running in GCE bound to a service account.  This VM may be used to run the integration tests without transferring a private key, increasing security.

Credentials Initialization
===

The developer tools image initializes credentials from the `SERVICE_ACCOUNT_JSON` input.  This initialization should not be automatic (e.g. initialization should not always run via the container's ENTRYPOINT) so that different jobs may customize the behavior if necessary.

The expected workflow for a developer working on their local workstation is:

 1. Enter the developer-tools container to a bash shell
 2. Load a well-known bash script as a library into the shell context
 3. Initialize credentials with `init_credentials`

Credentials initialization may require writing the key to the filesystem.  In this situation, a secure temporary file should be used and should be automatically cleaned up when the running container reaches end of life, or sooner depending on the use case.

Example Implementation
===

An example implementation is provided in #250 and the local development workflow looks like:

Enter the container:

```
docker run --rm -e SERVICE_ACCOUNT_JSON -it \
  -v `pwd`:/workspace \
  cft/developer-tools:0.0.1
```

Load the helper functions:

```
source /usr/local/bin/task_helper_functions.sh
```

Initialize credentials:

```
init_credentials
```

At this point, gcloud, gsutil, inspec and terraform are all expected to work without having to specify a credentials file to load from disk.  See the `init_credentials()` function for the distinct initialization steps.

Per-module customization
===

A module may need to customize the behavior of one of the task helper functions, e.g. `run_integration_tests`.  The module should create a `test/task_helper_functions.sh` file which defines the function to be overridden, e.g. `run_integration_tests() { ... }`.

`/usr/local/bin/task_helper_functions.sh` loads `/workspace/test/task_helper_functions.sh` after defining the common function.  This provides the mechanism to override these common functions.

**Comment by jeffmccune** [comment]
Once #250 is merged, I suggest resolving this ticket.  The ticket is intended to capture design decisions from recent conversations about the developer experience.

**Comment by morgante** [comment]
One clarification in the expected workflow: the end developer should *not* have to load the helper functions or run init_containers. When they enter the container, those functions should already have been run for them and they get a working authenticated bash shell.

**Comment by jeffmccune** [comment]
> When they enter the container, those functions should already have been run for them and they get a working authenticated bash shell.

Agree, this is nice to have.  What mechanism should provide this behavior?  `make docker_run`?

In order to avoid issues it's important to also have a mechnism to enter the container _without_ credentials being initialized.  For example, the use cases of running lint checks and documentation generation should not require credentials to be initialized.  It's also helpful to not execute `init_credentials` upon entry when troubleshooting or modifying the `init_credentials` function itself.

Using ENTRYPOINT to initialize credentials caused issues because it executed unconditionally and too soon to be skipped.

**Comment by jeffmccune** [comment]
@morgante FYI, I updated the [Credentials Initialization](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/issues/255#issue-473456849) section to state interactive shells automatically have credentials initialized, while non-interactive shells (e.g. `docker run ... bash -c make foo`) do not have credentials initialized.
